mavenize openHAB and integrate mavenized ESH repository

maggu2810 opened the pull request:
This commit integrates the whole mavenized ESH repository ATM.
I would prefer to drop the ESH extensions later, because IIRC there are some openHAB bundles that depends on an extrension.
Perhaps we could drop the ESH extensions as soon as we moved some openHAB non framework bundles to the addons repository.

maggu2810 commented:
I had to amend an additional change, so we can use the openHAB Oomph setup file after the merge has been done immediately (I hope I did not miss a change of it -- otherwise we can create additional PRs to fix it).

maggu2810 commented on NOTICE:
```diff
@@ -0,0 +1,194 @@
+This content is produced and maintained by the Eclipse SmartHome project.
```
What about the notice file in the root path -- there has been none for the current openHAB core repo.

kaikreuzer replied:
remove it

kaikreuzer replied:
The LICENSE file is enough there.

maggu2810 replied:
I will remove it.

kaikreuzer commented on bom/coap/pom.xml:
```diff
@@ -0,0 +1,42 @@
+  <artifactId>org.eclipse.smarthome.bom.coap</artifactId>
+  <packaging>pom</packaging>
+
+  <name>Eclipse SmartHome # BOM # COAP</name>
```
There's still "Eclipse SmartHome" in all the bom files.

kaikreuzer commented on bom/compile-model/.project:
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<projectDescription>
+	<name>org.eclipse.smarthome.bom.compile-model</name>
```
`org.eclipse.smarthome` -> `org.openhab`.
But did we actually want the bom projects to be in the workspace? If not, those files should be removed.

maggu2810 replied:
Sure, the BOMs provide the indices that are used by the launch configuration (bnd app).

maggu2810 commented:
For the naming fixes (I assume we will find a lot others) can (one of the alternatives)
* you and the other developers create PRs to the source branch of this PR
* this be merged and the fixes can be done by separate PRs against the master

otherwise it depends when I will find time to do it myself

kaikreuzer commented:
Sure, I'd suggest we do it as with any other PRs - comment on the PR and creating PRs against your branch with suggested changes until it is ready to be merged.

wborn commented:
Thanks for your huge efforts @maggu2810! :+1: 

I had a bit of trouble to build your branch but after wiping my local Maven repository the build succeeded.

It kept throwing the following compilation error:

```
[INFO] Compiling 3 source files to /home/wouter/git/openhab/openhab-core/bundles/io/org.eclipse.smarthome.io.http.auth/target/classes
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /home/wouter/git/openhab/openhab-core/bundles/io/org.eclipse.smarthome.io.http.auth/src/main/java/org/eclipse/smarthome/io/http/auth/internal/AuthenticationHandler.java:[110,21] cannot find symbol
  symbol:   method getStatus()
  location: variable response of type javax.servlet.http.HttpServletResponse
[ERROR] /home/wouter/git/openhab/openhab-core/bundles/io/org.eclipse.smarthome.io.http.auth/src/main/java/org/eclipse/smarthome/io/http/auth/internal/AuthenticationHandler.java:[110,52] cannot find symbol
  symbol:   method getStatus()
  location: variable response of type javax.servlet.http.HttpServletResponse
```

Any reasons for [downgrading Karaf to 4.2.1 (instead of 4.2.2)](https://github.com/maggu2810/openhab-core/blob/9ea5510e0dfcfd112841612d71b93b44c050613c/pom.xml#L59)? We just upgraded to Karaf 4.2.2 with https://github.com/openhab/openhab-core/pull/458. 

Because there are so many changes, the normal GitHub review UI seems unusable for adding comments. 
Does someone know of another way to add inline comments with such huge PRs?

maggu2810 commented:
> Any reasons for [downgrading Karaf to 4.2.1 (instead of 4.2.2)](https://github.com/maggu2810/openhab-core/blob/9ea5510e0dfcfd112841612d71b93b44c050613c/pom.xml#L59)? We just upgraded to Karaf 4.2.2 with #458.

Hi @wborn,
I don't see any problem here.

openHAB Core has been moved to a framework and I don't see any reason why org.openhab.core.karaf shouldn't compile against Karaf 4.2.1.
I assume we could also depend on 4.1.1 or something similar.

Or do you use any feature that has not been present before?

You should differ between version that we compile against and the version that will be used in the product.

I don't think that any usage of Karaf 4.2.1 for the framework will break your Karaf 4.2.2 usage in the product.

But to answer your question: No, there has not been any reason. I assume it has been set by me on the migration path and I don't bumped it because I have not identified a problem.

maggu2810 commented:
@kaikreuzer 

1. I don't think we need should use a `pom` of openHAB Core as parent `pom` for the addons or the distribution. Perhaps a migration to `openhab-infrastructure` would make sense. The framework itself should not contain stuff that is not relevant for the framework.

2. The BSN and automatic module name currently uses the artifact ID. The artifact ID has not been changed yet, because this would result into some collisions. There is now an "openHAB Core" and an "Eclipse SmartHome Core" bundle. There are UI specific Eclipse SmartHome bundles (Classic, Basic, Paper) and the fragment / patch bundle of openHAB. If we want to change the artifact ID and the BSN already, we need to move the UI bundles to openHAB Addons first (you stated a PR has been created for the ESH ones, but we need to also integrate the openHAB ones after that into the merged addons). Also the openHAB Core needs to be migrated into Eclipse SmartHome (and then renamed). We could also move the non framework relevant openHAB Core bundles in front of this one (I don't know if the "start" bundle is really a framework one and so on).

kaikreuzer commented:
1. Well, the poms serve as a parent pom, which provide configuration for required external repos, SAT checks, etc. So imho it is closely related to the core framework as it is the required info for "how to build add-ons for the framework". But I am fine with moving this to the infrastructure project (@pfink, I assume that's ok with you as well).

2. As discussed, I am already on it (removing the addon bundles from core), see https://github.com/openhab/openhab-core/pull/469. So in your PR, you can directly remove those as well and the conflicts should be gone.

maggu2810 commented:
1. That's perhaps true for addons (e.g. bindings) that are provided by openHAB itself. If another one would like to develop addons for the framework (that are not under control of openHAB itself), he need to overwrite all openHAB related stuff that is part of the parent POM. So he needs to know the whole parent pom if there is stuff insert that needs be overwritten. If the POM it not intended to be used by non-openHAB owned bundles I don't think it should be part of this repo.

2. Remaining conflict: `org.openhab.core` and `org.eclipse.smarthome.core` (that will be `org.openhab.core` after renaming)

maggu2810 commented:
After that we also need to merge the Karaf feature repositories.
I assume the prefix `esh-` should be renamed to `openhab-`. As I don't know anything about the openHAB features (which one exists), do you see here any conflicts?
Correct?

After the work has been done there will be two feature repos:
* openhab-tp: the old esh-tp
* openhab-core: the merge of `esh-core` and `openhab-core`

wborn commented:
> Or do you use any feature that has not been present before?

It's OK to have the framework on another version as long as there is another way (like Kai already suggested) to manage the Karaf/Jetty versions that are used in the distro. 

Without the proper Karaf/Jetty ending up in the distro we'll have regressions like the [Karaf SSH Console new line bug](https://github.com/openhab/openhab-distro/issues/764) and [Jetty ThreadPoolBudget warnings](https://community.openhab.org/t/low-configured-threads-warning/52084?u=wborn).

maggu2810 commented:
That's a very important point, the framework must not dictate the exact version of Karaf and Jetty that needs to be used by a product.
I don't see any change that e.g. openHAB and the products I am maintaining (and perhaps other consumers of the framework) will use the same versions of Karaf and Jetty all the time.

kaikreuzer commented:
> I assume the prefix esh- should be renamed to openhab-. As I don't know anything about the openHAB features (which one exists), do you see here any conflicts?

I am wondering whether we should maybe even use `openhab-core-` or `openhabcore-` as a prefix for all of them, so that it is easier to determine where they belong to / where to find them?

> After the work has been done there will be two feature repos

Sounds right.

maggu2810 commented:
> > I assume the prefix esh- should be renamed to openhab-. As I don't know anything about the openHAB features (which one exists), do you see here any conflicts?
> 
> I am wondering whether we should maybe even use `openhab-core-` or `openhabcore-` as a prefix for all of them, so that it is easier to determine where they belong to / where to find them?

I already think about it, but I did not yet know which is the best approach.
The consumer of features does not care which one provides it.

If you have a look at the whole Karaf features, you don't care which one provides it. The name should state what will be provided by the feature.

I start thinking about the whole thing as it has been about the group ID.
Should all core / framework bundles start with a group ID prefix of `org.openhab.core` instead only `org.openhab`? Using `org.openhab` and not `org.openhab.core` we don't know if a bundle that uses the group ID `org.openhab.ui` is part of the framework / core or if it is "some addon".

So, I am not against the suggested feature prefix, but if we agree, we have to change the group ID, too. Otherwise our argumentation is not really consistent.

maggu2810 commented:
The current openHAB Core bundles depend on the models, too.
Will this bundle be moved to the distribution or the addons, too?

kaikreuzer commented:
> Should all core / framework bundles start with a group ID prefix of org.openhab.core instead only org.openhab?

It imho makes sense to be able to differentiate between framework and add-on through the group ids. In ESH we had "esh.extension.*" for the addons, but I think using "org.openhab.core" for all framework parts is the better choice.

> The current openHAB Core bundles depend on the models, too.

You mean the Xtext bundles? Yes, they are an integral part of the framework, just like the config/io/automation bundles as well.

maggu2810 commented:
Another try... It seems my question has been not clear enough, so let me add more details:

There is a bundle with the name "openHAB Core" that uses the group ID `org.openhab.core` and the artifact ID `org.openhab.core` (https://github.com/openhab/openhab-core/blob/f9394266f2a572b684d2339cc6425bcc08c6a0f5/bundles/org.openhab.core/pom.xml).

There is a bundle with the name "Eclipse SmartHome Core" that uses the group ID `org.eclipse.smarthome.core` and the artifact ID `org.eclipse.smarthome.core` (https://github.com/eclipse/smarthome/blob/2239635027bf1e809b97dc3292f6da982e5daa0e/bundles/core/org.eclipse.smarthome.core/pom.xml).

If we merge the core bundles of openHAB and Eclipse Smarthome and replace "Eclipse SmartHome" with "openHAB" and "org.openhab.core" with "org.eclipse.smarthome.core", the bundles collide.

The esh.core bundle contains the real core and is referenced (transitive) by the most of the other bundles.

The oh.core bundle (correct me if I am wrong) is not really a core part of the whole smarthome framework. A product works perfect without it.

But okay, we change a lot of things now so perhaps both bundles should be merged into one. If all can coexist why should we care about.

The oh.core bundle imports model stuff (see `https://github.com/openhab/openhab-core/blob/f9394266f2a572b684d2339cc6425bcc08c6a0f5/bundles/org.openhab.core/META-INF/MANIFEST.MF#L36) I don't think this will result in a cycle or don't esh.model depend on esh.core itself.

> You mean the Xtext bundles? Yes, they are an integral part of the framework, just like the config/io/automation bundles as well.

Sure, like config, IO and automation, it is part of the core but we can also create a product without the automation stuff.

I would veto if the most core bundle (what should be named "openHAB Core" similar to "Eclipse SmartHome Core" before) adds a hard dependency to the Xtext bundles.
Adding a requirement for Xtext to the base bundle of the framework is IMHO a really hard break to all consumers of the framework.

maggu2810 commented:
About the group and artifact IDs to use, can you please give me a example?

The current approach has been to replace `org.eclipse.smarthome` with `org.openhab`.
So there will be `org.openhab.core`, `org.openhab.ui`, `org.openhab.auth`, `org.openhab.io`, ... groups that contains bundles.
If we replace with `org.openhab.core` it is fine for `org.openhab.core.ui`, `org.openhab.core.auth`, `org.openhab.core.io`, ... but I don't think we would like `org.openhab.core.core`.

kaikreuzer commented:
> If we merge the core bundles of openHAB and Eclipse Smarthome and replace "Eclipse SmartHome" with "openHAB" and "org.openhab.core" with "org.eclipse.smarthome.core", the bundles collide.

Ok, sorry, I didn't see you were talking about those - yes, indeed, they are a collision (hopefully the only one).
Looking at the content of openhab.core, I'd think it'd make sense to actually get rid off the bundle completely. The DefaultSitemapProvider is of hardly any use (not sure anybody uses it...), the IconFormwarder should probably better be part of the compat1x bundle and the activator stuff could be discussed to be useful in the framework (e.g. for the LSP fix). Maybe the easiest solution for now would be to simply rename this bundle to openhab.boot or something and thus avoid the collision.

> Adding a requirement for Xtext to the base bundle of the framework is IMHO a really hard break to all consumers of the framework.

That's clearly not wanted by anybody. The Xtext functionality is kept separate and must be kept this way.

> but I don't think we would like org.openhab.core.core

... while it would be the most logical name. It is the core part of the core framework. I wouldn't mind that group id.

kaikreuzer commented:
> ... while it would be the most logical name. It is the core part of the core framework. I wouldn't mind that group id.

Sleeping over it and noticing that this will mean on the long run that those are also the package names, I question this statement myself.
Probably it is better to use `org.openhab.core` for all other parts (automation/config/io) and avoid the duplicate `core` for the "core" bundles then. Afaics, we won't have any collisions apart from the one we have already discussed.

maggu2810 commented:
So you suggest to use the group ID `org.openhab.core` for all stuff in the openhab-core repository without any sublevels? One group ID for every artifact?
This should be possible because our artifact IDs already are unique around the whole directories.

Should we place all bundles directly in the directory without any further subdirectory between? Do we really need a directory `ui`, `auth`, `automation`, ...?

Please could you provide the group and artifact ID + directory layout for some bundles, so we could not misunderstand each other?

kaikreuzer commented:
> So you suggest to use the group ID org.openhab.core for all stuff in the openhab-core repository without any sublevels? One group ID for every artifact?

No, I thought we will have group ids `org.openhab.core.ui`, `org.openhab.core.auth`, etc. and then instead of `org.openhab.core.core` simply `org.openhab.core`.

> Should we place all bundles directly in the directory without any further subdirectory between? Do we really need a directory ui, auth, automation, ...?

I quite like the hierarchy instead of a very long flat list as it makes browsing a bit easier. So I wouldn't mind keeping it exactly as it is in your repo right now. If you think things get easier by having just the "bundles" folder and all bundles in it, it would be fine for me as well.

Just three examples, which should clarify be transferable to all the rest:

1. bundles/core/org.openhab.core: groupId=org.openhab.core, artifactId=org.openhab.core
1. bundles/core/org.openhab.core.voice: groupId=org.openhab.core, artifactId=org.openhab.core.voice
1. bundles/ui/org.openhab.core.ui.dashboard: groupId=org.openhab.core.ui, artifactId=org.openhab.core.ui.dashboard

Does that make sense?

pfink commented:
> I don't think we need should use a pom of openHAB Core as parent pom for the addons or the distribution. Perhaps a migration to openhab-infrastructure would make sense. The framework itself should not contain stuff that is not relevant for the framework.

> 1. Well, the poms serve as a parent pom, which provide configuration for required external repos, SAT checks, etc. So imho it is closely related to the core framework as it is the required info for "how to build add-ons for the framework". But I am fine with moving this to the infrastructure project (@pfink, I assume that's ok with you as well).

It would be good to carve out stuff that is not version-specific. openhab/infrastructure is designed to be a component that really contains only version-unrelated infrastructure informations, because it is (on purpose!) not versioned and tagged together with openHAB. This could be the information e.g. where the base Git repository url is located or which Maven repositories openHAB is currently using. 

But, what would be a very bad smell is to put version-specific information there (e.g. which versions of maven plugins are currently used).

> That's perhaps true for addons (e.g. bindings) that are provided by openHAB itself. If another one would like to develop addons for the framework (that are not under control of openHAB itself), he need to overwrite all openHAB related stuff that is part of the parent POM. So he needs to know the whole parent pom if there is stuff insert that needs be overwritten. If the POM it not intended to be used by non-openHAB owned bundles I don't think it should be part of this repo.

Sorry, I don't understand the whole point. Where the parent POM is located has IMO nothing to do with the question whether a developer has to inherit his addon from this parent POM or not.

maggu2810 commented:
> If openHAB decides one day to move to another SCM provider or to another artifact repository, this would affect all versions (also the older ones), so it's a not version-related change. In this case, of course it would be beneficial if the effort to make an older build work with the new environment is as low as possible. If we have the version-unrelated information at one place together, this effort would be minimal.

I don't understand how this will affect "older builds".
From a Maven perspective:
* releases builds must not depend on snapshots
* a release must never be changed

So, I assume the openHAB release "m" depends on the released infra version "x".
If the infra POM will be changed it needs another release of infra, so version "y".
openHAB (in its current development / master branch) can be changed to point to "y" now, but all versions of openHAB <= "m" will still rely on "x".

maggu2810 commented:
@kaikreuzer For the bundle part it make sense. Just a question about the parent POMs of the bundles. How do you want that the GA looks like?

Just to consider:

The reactor `./bundles/auth/pom.xml` should use a group ID of `org.openhab.core.auth` so we do not need to declare the group ID for every child again.
The reactor `./bundles/ui/pom.xml` should use a group ID of `org.openhab.core.ui` so we do not need to declare the group ID for every child again.
If every bundle below `./bundles/core` use a group ID of `org.openhab.core`, the reactor `./bundles/core/pom.xml` should use a group ID of `org.openhab.core`.

So, if `org.openhab.core` is used by `./bundles/core/pom.xml` we need to think about its parent POMs.

Example:

```
Format: directory/pom.xml: groupId/artifactId

* ./bundles/core/org.openhab.core/pom.xml: org.openhab.core/org.openhab.core
* ./bundles/core/org.openhab.core.voice/pom.xml: org.openhab.core/org.openhab.core.voice
* ./bundles/ui/org.openhab.core.ui.dashboard/pom.xml: org.openhab.core.ui/org.openhab.core.ui.dashboard

* ./bundles/auth/pom.xml: org.openhab.core.auth/org.openhab.core.reactor.bundles.auth.core
* ./bundles/ui/pom.xml: org.openhab.core.ui/org.openhab.core.reactor.bundles.ui.core
* ./bundles/core/pom.xml: org.openhab.core/org.openhab.core.reactor.bundles.core
* ./bundles/pom.xml: org.openhab.core.bundles/org.openhab.core.reactor.bundles
* ./pom.xml: org.openhab.core/org.openhab.core.reactor
```

Another option would be to consider about the name of the whole bundle collection.

Currently it is named "openHAB Core" perhaps it could also be the "openHAB Base", "openHAB Basis" or "openHAB Framework".

E.g. This could lead to a prefix of ... and its bundles of...

* `org.openhab.base`
  * `org.openhab.base.core`
  * `org.openhab.base.ui`

* `org.openhab.basis`
  * `org.openhab.basis.core`
  * `org.openhab.basis.ui`

* `org.openhab.framework`
  * `org.openhab.framework.core`
  * `org.openhab.framework.ui`

* `org.openhab.fw`
  * `org.openhab.fw.core`
  * `org.openhab.fw.ui`

kaikreuzer commented:
Maybe it would be simpler after all to have a flat folder for the bundles and simply use `org.openhab.core.bundles` as a groupId for its pom (next to `org.openhab.bom`, `org.openhab.feature` etc.)?

maggu2810 commented:
You mean next to `org.openhab.core.bom`, `org.openhab.core.features` (with a core in front of the subdir name)?

So, we could perhaps use `org.openhab.core.bundles` for the group ID of all bundles that resides in that repository.

If we would like to flatten the repository can be decided separated.
It depends if we want that the group ID fits exactly to the reactor directory layout or not.

e.g.

```
./pom: org.openhab.core/org.openhab.core.reactor
./bundles/pom.xml: org.openhab.core.bundles/org.openhab.core.reactor.bundles
./bundles/core/pom.xml: org.openhab.core.bundles/org.openhab.core.reactor.bundles.core
./bundles/core/org.openhab.core/pom.xml: org.openhab.core.bundles/org.openhab.core
./bundles/core/org.openhab.core.voice/pom.xml: org.openhab.core.bundles/org.openhab.core.voice
./bundles/ui/pom.xml: org.openhab.core.bundles/org.openhab.core.reactor.bundles.ui
./bundles/ui/org.openhab.core.ui.dashboard/pom.xml: org.openhab.core.bundles/org.openhab.core.ui.dashboard

or

./pom: org.openhab.core/org.openhab.core.reactor
./bundles/pom.xml: org.openhab.core.bundles/org.openhab.core.reactor.bundles
./bundles/org.openhab.core/pom.xml: org.openhab.core.bundles/org.openhab.core
./bundles/org.openhab.core.voice/pom.xml: org.openhab.core.bundles/org.openhab.core.voice
./bundles/org.openhab.core.ui.dashboard/pom.xml: org.openhab.core.bundles/org.openhab.core.ui.dashboard
```

Using separate directories will group the core parts like "org.openhab.core and "org.openhab.core.voice" separated from "org.openhab.core.ui.dashboard".
Reading just the names does not make clear that "org.openhab.core.voice" is part of "core" and not a separate group "voice".

kaikreuzer commented:
> So, we could perhaps use org.openhab.core.bundles for the group ID of all bundles that resides in that repository.

Would imho make sense.

> Reading just the names does not make clear that "org.openhab.core.voice" is part of "core" and not a separate group "voice".

Would be an argument for the flat structure, but I agree that this can also be changed later. So anything for which there is no sub-folder will naturally be found in the "core" folder then (such as the core.voice bundle).

maggu2810 commented:
> > Reading just the names does not make clear that "org.openhab.core.voice" is part of "core" and not a separate group "voice".
> 
> Would be an argument for the flat structure, but I agree that this can also be changed later. So anything for which there is no sub-folder will naturally be found in the "core" folder then (such as the core.voice bundle).

Is it an argument for a flat structure? I assume flat means all in one directory under bundles and keeping separate subdirectories is a hierarchical structure.
If we would like to make it more visible that `org.openhab.core.voice` is part of `core` and `org.openhab.core.ui.dashboard` is part of `ui`, we should use the subdirectories `core` and `ui`, so we could place `...core.voice` in `core` and `...core.ui.dashboard` in `ui`.

kaikreuzer commented:
Well, that's how it is right now: https://github.com/openhab/openhab-core/tree/master/bundles

`org.openhab.core.karaf` is right besides `org.openhab.ui.dashboard`.

> ... is part of `core` ...

All bundles are "part of core", just some of them also have the artifact/bundle id starting with `org.openhab.core`...

maggu2810 commented:
My assumption has been that you still want to change the name of the projects and the artifact IDs to contain "core" as prefix for all bundles.

So previously you have `org.openhab.core.karaf` and `org.openhab.ui.dashboard` that signals that the one is category "core" and the other category "ui" by reading the third part.

If we use `org.openhab.core.karaf` and `org.openhab.core.ui.dashbard` this is not the case anymore. For dashboard the "category" is at position four, for karaf it is the non-duplicated "core" and so on position three.

So, if I get you right, you don't want to add "core" on the project ID and artifact ID of `org.openhab.ui.dashboard` (between "ui" and "openhab") and it should stay as before?

If a developer mix addons and framework some time in its IDE the project name does not make visible if it belongs to the framework or not.
Okay for me, but then we don't need a "core" in the feature names.

kaikreuzer commented:
My point is that I do not see the need for a "category" for the core framework really. Why should "voice" be in the "core" category, while "ui" isn't?

> So, if I get you right, you don't want to add "core" on the project ID and artifact ID of org.openhab.ui.dashboard (between "ui" and "openhab") and it should stay as before?

No, I only pointed at the existing repo with my example. It perfectly makes sense to use `org.openhab.core` for all artifact/bundle IDs, i.e. `org.openhab.core.ui.dashboard`.

maggu2810 commented:
The refactoring of the bundle has been "some" work.
If someone wants another group ID, artifact ID or directory layout please create a PR. :wink: 

The compilation works except the feature (verification).
Will fix this on another day.

maggu2810 commented:
For the moment I would suggest to close the following projects in the IDE:

* ~~org.openhab.core.bom.runtime-index~~
* ~~org.openhab.core.bom.test-index~~
* org.openhab.core.tests
* org.openhab.core.compat1x.tests
* org.openhab.core.demo.app + org.openhab.core.demo.index

~~The BOM index projects create an index repo to be consumed by the demo application and tests.~~
The both tests are already converted tests.
The demo.app is the "launch config" for the IDE and the demo.index is an index for that one (has been split because my assumption has been it improved the thing but it seems it makes no difference).


The problem is IMHO this part of code: https://github.com/bndtools/bndtools/blob/master/bndtools.m2e/src/bndtools/m2e/IndexConfigurator.java#L343

This triggers a rebuild of the indices if a dependent project has been changed / rebuild.
The tests and the demo application will bill built after each bundle on a "rebuild all" and other changes. This slows down the whole thing.

@kaikreuzer perhaps you can use your contacts to the Bndtools dev team members some time (I don't think it blocks something but would be good if we don't need to close that projects and open it on request / usage).

maggu2810 commented:
Have you had a look at if it is ready to get merged?
Further fixes could perhaps be made later and we could continue now with further fixes of the framework code base.

maggu2810 commented:
Is there any other usecase for the "NOTICE" file then to be placed into the JAR (at its root)?
Is there a required location in the source tree?
I would lile to move that file to `src/main/resources` (the default resource directory) and drop the confoguration that needs to be added to the POM

kaikreuzer commented:
Thanks for all your work @maggu2810!

Just going through some the repo and create some small clean-up PRs.

A few things are missing, just want to check if this is correct:
- The parent poms as discussed above. If I get @pfink correctly, he does not feel that openhab-infrastructure would be the right place for them, so I am not sure, where to put them. I think the simplest temporary solution would be to keep them in this repo here as before and they will go away anyhow once we refactor the addons repos to bnd. Would that be ok?
- The `tools/archetype` is missing - I think it should be migrated from ESH to here just as it is.
- The p2 repo is missing, which we will need for the addons-IDE setup. If you like, I can take care of this job (while it would have been easier if you hadn't completely removed it in this PR).

Anything else that got lost on the way and that I haven't noticed yet?

maggu2810 commented:
Is there any other usecase for the "NOTICE" file then to be placed into the JAR (at its root)?
Is there a required location in the source tree?
I would lile to move that file to `src/main/resources` (the default resource directory) and drop the confoguration that needs to be added to the POM

kaikreuzer commented:
> Is there any other usecase for the "NOTICE" file then to be placed into the JAR (at its root)?

Well, the other use case is to be in the root of the sourcecode folder in the repo, so that it is easy to see&find. In `src/main/resources` it will be rather hidden. But I agree that from a Maven pov it makes sense to move it, so that the build has few exceptions to the default. So for me ok to move it.

maggu2810 commented:
Normally (IMHO) the code that is placed in a Git repository is published under one license. One license for the whole repository. Sure there are some exceptions but one license is (perhaps personal feeling) the common case.
The license of the stuff placed in a repo is mostly found in a file called "LICENSE{,.md,.txt)" that resides in the root directory of the repository.
I normally don't dig into some sub-directories to look if there is some code that is using another license.
It differs for the build products, all the bundles etc. Here I agree that it should be in the root of the JAR.

So, perhaps we could add a license file to the root of the repository that states that the stuff in that repo is using EPL 2.0 and move the NOTICE file for the bundles into `src/main/resources`.

Okay?

maggu2810 commented:
>     * The parent poms as discussed above. If I get @pfink correctly, he does not feel that openhab-infrastructure would be the right place for them, so I am not sure, where to put them. I think the simplest temporary solution would be to keep them in this repo here as before and they will go away anyhow once we refactor the addons repos to bnd. Would that be ok?

I don't understand its whole comment and so I replied to it and wait for a reply myself.
Perhaps he could also create a PR what information needs to be shared and we can talk / comment that one.

>     * The `tools/archetype` is missing - I think it should be migrated from ESH to here just as it is.

As stated in notes/nodes.md - I know. ;)
Also SAT needs to be enabled again.

>     * The p2 repo is missing, which we will need for the addons-IDE setup. If you like, I can take care of this job (while it would have been easier if you hadn't completely removed it in this PR).

There is also a small comment in the notes file.
As we should not rely on most of the previously used maven plugins and custom packaging, I added a link to a page that explains how it can be done.
If you want to use the old one, the removal should not really be the problem. Is it? Just copy the files of the current master branch to your local working copy (point to that branch) and add it again.

kaikreuzer commented:
> So, perhaps we could add a license file to the root of the repository

We already have that in place in all our repos.
The speciality about the NOTICE file in the bundle folders is that they contain the exceptional license informations, e.g. about included 3rd party libs. This belongs to the bundles (and should not be in the repo root) and will be hidden if we move the files to `src/main/resources`. But still: Ok for me.

maggu2810 commented:
> > So, perhaps we could add a license file to the root of the repository
> 
> We already have that in place in all our repos.

I must be blind. Sorry

marziman commented:
Hi @maggu2810,

we are running our runtime of ESH on Eclipse Concierge. It was mainly based on the ESH sample packaging back than: https://github.com/eclipse/smarthome-packaging-sample/blob/master/README.md

So we use also a mavenized approach.
We would like to early test that the transfer of ESH framework to openHAB against our distro packaging.

Whats the best way? WDYT? Do you see any blockers? For the time being I tought that all bundles remain modular and we just will face namespace changes and the modification of our maven build towards the OH-Framework jars.

Thanks for efforts.
BR Mehmet

kaikreuzer commented:
@marziman This PR is already close to getting merged. The output of it will be a set of bundles that is almost identical to the bundles from ESH (just with different Maven GAVs and bundle IDs), so there should absolutely be no issue to package it similar to the ESH sample packaging.
Just pull this branch, build it locally and go ahead.

pfink commented:
> openHAB (in its current development / master branch) can be changed to point to "y" now, but all versions of openHAB <= "m" will still rely on "x".

Yes, that's true, there is no magic that would let it work automatically. But, instead of having to search, compare & replace all relevant occurences manually, you'd just have to adjust the Infra version at a single place.

Hilbrand commented:
What about the still open points, like maven archetype and SAT? Or merge first and fix later?

kaikreuzer commented:
> Or merge first and fix later?

That would be my plan.
Any volunteer for adding the maven archetype? Should probably be rather straight forward.
@maggu2810 Wrt SAT, you might be the best to tell what problems you came across and what might be the best way to include it - so your input is very welcome!

maggu2810 commented:
> @maggu2810 Wrt SAT, you might be the best to tell what problems you came across and what might be the best way to include it - so your input is very welcome!

I didn't test it yet because of the lack of time.
Perhaps enabling it again would be enough, but I assume as we moved ESH-INF there need to be some changes. I don't know which paths are hardcoded.

But I don't think it is a blocker and can be integrated also later.

maggu2810 commented:
@kaikreuzer As long as there is no solution for https://github.com/openhab/openhab-core/pull/467#issuecomment-456362798, should we remove the `.project` files for that three projects, so they are not added to the projects after someone executed the IDE setup.

kaikreuzer commented:
Can you do a PR?

maggu2810 commented:
Yes, i will do.

FYI: Without the index projects I still realized an interesting behavior of the IDE.
Interesting in that case that I don't have an explanation for it yet.
Yesterday on one of my machines I realized that there seems to be an endless loop building "org.openhab.core" (inspecting the "Progress" view of the Eclipse IDE).
I didn't realized it before and reflect what has been changed.
The only change I could identify is that I switched from "Oracle JDK" to "Adopt OpenJDK" to run the IDE.
On another machine running Oracle JDK 8 as IDE VM I did not realize that behavior.
I changed the other machine from "Adopt OpenJDK" to  "Oracle JDK" 8 again and it seems to be solved.

I don't know if the specific Java distribution is really the root cause of it but we should remember, if someone complains.

Hilbrand commented:
@maggu2810 I did 'complain' about this already twice, see pure maven thread. So finally I don't feel alone anymore :slightly_smiling_face: I run:

    openjdk version "1.8.0_191"
    OpenJDK Runtime Environment (build 1.8.0_191-8u191-b12-0ubuntu0.18.04.1-b12)
    OpenJDK 64-Bit Server VM (build 25.191-b12, mixed mode)

It makes development completely impossible, and why I, kind of, postponed looking into maven-only-build until some time later in the hope it was just because something temporary due to the wip status and nobody else seem to have that problem.

So it looks like specific Java distribution problem...

maggu2810 commented:
> and why I, kind of, postponed looking into maven-only-build until some time later in the hope it was just because something temporary due to the wip status and nobody else seem to have that problem.

Hm, isn't one of the pros of the maven only build that you don't rely on the Eclipse IDE (sure, for the formatter you still need to use the Eclipse IDE) but for "looking into" you could use every IDE that can handle Maven projects.

> So it looks like specific Java distribution problem...

Please try again using the Oracle JDK 8 (202 should be the recent one).

Hilbrand commented on pom.xml:
```diff
@@ -59,29 +48,483 @@
+        <artifactId>org.openhab.core.bom.compile</artifactId>
+        <version>${project.version}</version>
+        <type>pom</type>
+        <scope>provided</scope>
```
I needed to change the scope to `compile` to be able to build it with maven on the command line. Same for provided below on `compile-model`

maggu2810 replied:
Really?
I build the whole repository multiple times now and it works.

If the build does not work for you (using the current master), it would perhaps make sense to create an issue with the whole log.

`provided` should be present on compile time (similar to `compile`) but can be missing at `runtime` (different to `compile`).

Hilbrand commented on pom.xml:
```diff
@@ -59,29 +48,483 @@
+      <url>http://eclipse.github.io/smarthome/third-party/m2-repo/</url>
+    </repository>
+    <repository>
+      <id>bintray-maggu2810-maven</id>
```
The url to the maggu2810 repository should be removed. Is something like that still needed?

maggu2810 replied:
It can be removed if you know about a repository that provides the dbus and unix bundle of matthw.
See: https://github.com/openhab/openhab-core/blob/c0ef794924d217d546da29eeda9a739d2cf349b0/bundles/org.openhab.core.io.transport.dbus/pom.xml#L16-L27

Hilbrand commented:
I manually installed openjdk 8 version 201 and that seem to solve the continuous build problem.
